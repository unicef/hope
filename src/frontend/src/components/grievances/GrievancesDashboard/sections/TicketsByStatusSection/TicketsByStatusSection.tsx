import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { AllGrievanceDashboardChartsQuery } from '@generated/graphql';
import { TicketsByStatusChart } from '../../charts/TicketsByStatusChart';
import { DashboardPaper } from '../../DashboardPaper';

interface TicketsByStatusSectionProps {
  data: AllGrievanceDashboardChartsQuery['ticketsByStatus'];
}

export function TicketsByStatusSection({
  data,
}: TicketsByStatusSectionProps): React.ReactElement {
  const { t } = useTranslation();
  return (
    <DashboardPaper noMarginTop title={t('Tickets by Status')}>
      <TicketsByStatusChart data={data} />
    </DashboardPaper>
  );
}
