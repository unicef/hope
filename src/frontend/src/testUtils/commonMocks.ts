import { vi } from 'vitest';
import React from 'react';

/**
 * Common mock for @hooks/useBaseUrl hook
 */
export const mockUseBaseUrl = (
  businessArea = 'afghanistan',
  programId = 'test-program',
  baseUrl = 'afghanistan/test-program',
) => {
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({ businessArea, programId, baseUrl }),
  }));
};

/**
 * Common mock for useProgramContext hook
 */
export const mockUseProgramContext = (
  groupLabel = 'Household',
  groupLabelPlural = 'Households',
) => {
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel,
          groupLabelPlural,
        },
      },
    }),
  }));
};

/**
 * Common mock for react-router-dom
 */
export const mockReactRouterDom = () => {
  vi.mock('react-router-dom', () => ({
    useNavigate: () => vi.fn(),
    useLocation: () => ({
      pathname: '/test-path',
      search: '',
      hash: '',
      state: null,
      key: 'test-key',
    }),
    Link: ({ children, to, ...props }: any) =>
      React.createElement('a', { href: to, ...props }, children),
  }));
};
/**
 * Common mock for @utils/utils module
 */
export const utilsMock = {
  createApiParams: vi.fn((baseParams, queryParams, options) => ({
    ...baseParams,
    ...queryParams,
    ...options,
  })),
  adjustHeadCells: vi.fn((headCells) => headCells),
  formatCurrencyWithSymbol: vi.fn(
    (amount, currency) => `${amount} ${currency}`,
  ),
  householdStatusToColor: vi.fn(() => '#000000'),
  filterEmptyParams: vi.fn((params) => {
    if (!params) return {};
    return Object.fromEntries(
      Object.entries(params).filter(([, value]) => {
        return value !== undefined && value !== null && value !== '';
      }),
    );
  }),
  opacityToHex: vi.fn((opacity) =>
    Math.round(opacity * 255)
      .toString(16)
      .padStart(2, '0'),
  ),
  isPermissionDeniedError: vi.fn((error) => error?.response?.status === 403),
  dateToIsoString: vi.fn((date) => date?.toISOString?.() || date),
  choicesToDict: vi.fn((choices) => {
    if (!choices) return {};
    return choices.reduce((acc, choice) => {
      acc[choice.value] = choice.name;
      return acc;
    }, {});
  }),
  populationStatusToColor: vi.fn(() => 'primary'),
  sexToCapitalize: vi.fn(
    (sex) => sex?.charAt(0).toUpperCase() + sex?.slice(1).toLowerCase(),
  ),
  programStatusToColor: vi.fn(() => 'primary'),
  formatCurrency: vi.fn((amount) => `$${amount?.toLocaleString() || '0'}`),
  columnToOrderBy: vi.fn((column) => column || 'name'),
};

/**
 * Helper function to mock the @utils/utils module
 */
export const mockUtils = () => {
  vi.mock('@utils/utils', () => utilsMock);
};

/**
 * Common mock for RestService
 */
export const mockRestService = () => {
  vi.mock('@restgenerated/services/RestService', () => ({
    RestService: {
      restBusinessAreasProgramsHouseholdsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsCountRetrieve: vi.fn(),
      restBusinessAreasProgramsIndividualsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsMembersList: vi.fn(),
      restBusinessAreasProgramsList: vi.fn(),
      restBusinessAreasProgramsCountRetrieve: vi.fn(),
      restBusinessAreasUsersProfileRetrieve: vi.fn(),
      // Add other commonly used RestService methods as needed
    },
  }));
};

/**
 * Helper function to set up all common mocks at once
 * Usage in test files:
 *
 * import { setupCommonMocks } from 'src/testUtils/commonMocks';
 *
 * // At the top level of your test file
 * setupCommonMocks();
 */
export const setupCommonMocks = () => {
  // Mock useBaseUrl
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({
      businessArea: 'afghanistan',
      programId: 'test-program',
      baseUrl: 'afghanistan/test-program',
    }),
  }));

  // Mock useProgramContext
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel: 'Household',
          groupLabelPlural: 'Households',
        },
      },
    }),
  }));

  mockReactRouterDom();
  mockUtils();
  mockRestService();
};
