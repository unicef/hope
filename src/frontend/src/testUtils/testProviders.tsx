import { ThemeProvider } from '@mui/material';
import CssBaseline from '@mui/material/CssBaseline';
import * as React from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { theme } from '../theme';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';

interface TestProvidersProps {
  children: React.ReactNode;
}

export const TestProviders: React.FC<TestProvidersProps> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <StyledThemeProvider theme={theme}>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <CssBaseline />
        {children}
      </LocalizationProvider>
    </StyledThemeProvider>
  </ThemeProvider>
);
