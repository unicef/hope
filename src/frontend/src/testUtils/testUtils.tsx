import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
