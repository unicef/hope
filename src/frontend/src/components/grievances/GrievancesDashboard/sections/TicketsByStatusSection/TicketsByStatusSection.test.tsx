import * as React from 'react';
import { render } from '../../../../../testUtils/testUtils';
import { fakeChartTicketsByStatus } from '../../../../../../fixtures/grievances/grievancesDashboard/fakeChartTicketsByStatus';
import { TicketsByStatusSection } from './TicketsByStatusSection';

describe('components/grievances/GrievancesDashboard/TicketsByStatusSection', () => {
  it('should render', () => {
    const { container } = render(
      <TicketsByStatusSection data={fakeChartTicketsByStatus} />,
    );
    expect(container).toMatchSnapshot();
  });
});
