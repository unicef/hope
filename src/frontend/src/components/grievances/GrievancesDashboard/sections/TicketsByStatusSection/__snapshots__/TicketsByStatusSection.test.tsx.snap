// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/grievances/GrievancesDashboard/TicketsByStatusSection should render 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 sc-blHHSb bfMGrt css-1ps6pg7-MuiPaper-root"
  >
    <div
      class="MuiBox-root css-i3pbo"
    >
      Tickets by Status
    </div>
    <div
      class="MuiBox-root css-1xrvl2k"
    >
      <canvas
        height="0"
        role="img"
        style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
        width="0"
      />
    </div>
  </div>
</div>
`;
